cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -0,0 +1,359 @@
+          level_val = 0;
+          if (cur_thread < end) { level_val = cur_thread[0]; }
+          cur_thread++;
+          if (level_bits > 8 - bitpos && cur_thread < end) {
+            level_val |= cur_thread[0] << 8;
+            cur_thread++;
+            if (level_bits > 16 - bitpos && cur_thread < end) { level_val |= cur_thread[0] << 16; }
+          }
+          level_val = (level_val >> bitpos) & ((1 << level_bits) - 1);
+        }
+
+        cur += batch_len8 * level_bits;
+      }
+
+      // store level_val
+      if (lane < batch_len && (lane + output_pos) >= 0) { output[lane + output_pos] = level_val; }
+      remain -= batch_len;
+      output_pos += batch_len;
+    }
+  }
+};
+
+// a single rle run. may be broken up into multiple rle_batches
+template <typename level_t>
+struct rle_run {
+  int size;  // total size of the run
+  int output_pos;
+  uint8_t const* start;
+  int level_run;  // level_run header value
+  int remaining;
+
+  __device__ __inline__ rle_batch<level_t> next_batch(level_t* const output, int max_size)
+  {
+    int const batch_len  = min(max_size, remaining);
+    int const run_offset = size - remaining;
+    remaining -= batch_len;
+    return rle_batch<level_t>{start, run_offset, output, level_run, batch_len};
+  }
+};
+
+// a stream of rle_runs
+template <typename level_t>
+struct rle_stream {
+  int level_bits;
+  uint8_t const* start;
+  uint8_t const* cur;
+  uint8_t const* end;
+
+  int max_output_values;
+  int total_values;
+  int cur_values;
+
+  level_t* output;
+
+  rle_run<level_t>* runs;
+  int run_index;
+  int run_count;
+  int output_pos;
+  bool spill;
+
+  int next_batch_run_start;
+  int next_batch_run_count;
+
+  __device__ rle_stream(rle_run<level_t>* _runs) : runs(_runs) {}
+
+  __device__ void init(int _level_bits,
+                       uint8_t const* _start,
+                       uint8_t const* _end,
+                       int _max_output_values,
+                       level_t* _output,
+                       int _total_values)
+  {
+    level_bits = _level_bits;
+    start      = _start;
+    cur        = _start;
+    end        = _end;
+
+    max_output_values = _max_output_values;
+    output            = _output;
+
+    run_index            = 0;
+    run_count            = 0;
+    output_pos           = 0;
+    spill                = false;
+    next_batch_run_start = 0;
+    next_batch_run_count = 0;
+
+    total_values = _total_values;
+    cur_values   = 0;
+  }
+
+  __device__ inline thrust::pair<int, int> get_run_batch()
+  {
+    return {next_batch_run_start, next_batch_run_count};
+  }
+
+  // fill in up to num_rle_stream_decode_warps runs or until we reach the max_count limit.
+  // this function is the critical hotspot.  please be very careful altering it.
+  __device__ inline void fill_run_batch(int max_count)
+  {
+    // if we spilled over, we've already got a run at the beginning
+    next_batch_run_start = spill ? run_index - 1 : run_index;
```

Comment:
Should we be using atomics to access/modify `next_batch_run_start` and `next_batch_run_count`?